File: awesome-skills-claude-agents/frontend/src/services/agents.ts
import api from './api';
import type { Agent, AgentCreateRequest, AgentUpdateRequest } from '../types';

// Convert camelCase to snake_case for API requests
const toSnakeCase = (data: AgentCreateRequest | AgentUpdateRequest) => {
  const result: Record<string, unknown> = {};
  if (data.name !== undefined) result.name = data.name;
  if (data.description !== undefined) result.description = data.description;
  if (data.model !== undefined) result.model = data.model;
  if (data.permissionMode !== undefined) result.permission_mode = data.permissionMode;
  if (data.systemPrompt !== undefined) result.system_prompt = data.systemPrompt;
  if (data.skillIds !== undefined) result.skill_ids = data.skillIds;
  if (data.mcpIds !== undefined) result.mcp_ids = data.mcpIds;
  if (data.allowedTools !== undefined) result.allowed_tools = data.allowedTools;
  if (data.enableBashTool !== undefined) result.enable_bash_tool = data.enableBashTool;
  if (data.enableFileTools !== undefined) result.enable_file_tools = data.enableFileTools;
  if (data.enableWebTools !== undefined) result.enable_web_tools = data.enableWebTools;
  return result;
};

// Convert snake_case response to camelCase
const toCamelCase = (data: Record<string, unknown>): Agent => {
  return {
    id: data.id as string,
    name: data.name as string,
    description: data.description as string | undefined,
    model: data.model as string | undefined,
    permissionMode: data.permission_mode as Agent['permissionMode'],
    systemPrompt: data.system_prompt as string | undefined,
    allowedTools: (data.allowed_tools as string[]) || [],
    skillIds: (data.skill_ids as string[]) || [],
    mcpIds: (data.mcp_ids as string[]) || [],
    workingDirectory: data.working_directory as string,
    enableBashTool: data.enable_bash_tool as boolean,
    enableFileTools: data.enable_file_tools as boolean,
    enableWebTools: data.enable_web_tools as boolean,
    enableToolLogging: data.enable_tool_logging as boolean,
    enableSafetyChecks: data.enable_safety_checks as boolean,
    status: data.status as 'active' | 'inactive',
    createdAt: data.created_at as string,
    updatedAt: data.updated_at as string,
  };
};

export const agentsService = {
  // List all agents
  async list(): Promise<Agent[]> {
    const response = await api.get<Record<string, unknown>[]>('/agents');
    return response.data.map(toCamelCase);
  },

  // Get agent by ID
  async get(id: string): Promise<Agent> {
    const response = await api.get<Record<string, unknown>>(`/agents/${id}`);
    return toCamelCase(response.data);
  },

  // Create new agent
  async create(data: AgentCreateRequest): Promise<Agent> {
    const response = await api.post<Record<string, unknown>>('/agents', toSnakeCase(data));
    return toCamelCase(response.data);
  },

  // Update agent
  async update(id: string, data: AgentUpdateRequest): Promise<Agent> {
    const response = await api.put<Record<string, unknown>>(`/agents/${id}`, toSnakeCase(data));
    return toCamelCase(response.data);
  },

  // Delete agent
  async delete(id: string): Promise<void> {
    await api.delete(`/agents/${id}`);
  },

  // Get default agent
  async getDefault(): Promise<Agent> {
    const response = await api.get<Record<string, unknown>>('/agents/default');
    return toCamelCase(response.data);
  },

  // Get available models
  async listModels(): Promise<string[]> {
    const response = await api.get<string[]>('/agents/models');
    return response.data;
  },
};

File: awesome-skills-claude-agents/backend/routers/agents.py
"""Agent CRUD API endpoints."""
from fastapi import APIRouter
from schemas.agent import AgentCreateRequest, AgentUpdateRequest, AgentResponse
from database import db
from config import ANTHROPIC_TO_BEDROCK_MODEL_MAP
from core.exceptions import (
    AgentNotFoundException,
    ValidationException,
)

router = APIRouter()


@router.get("/models", response_model=list[str])
async def list_available_models():
    """List all available Claude models.

    Returns the Anthropic model IDs that have Bedrock mappings configured.
    """
    return list(ANTHROPIC_TO_BEDROCK_MODEL_MAP.keys())


@router.get("", response_model=list[AgentResponse])
async def list_agents():
    """List all agents."""
    agents = await db.agents.list()
    # Filter out the default agent from the list
    return [a for a in agents if a.get("id") != "default"]


@router.get("/default", response_model=AgentResponse)
async def get_default_agent():
    """Get the default system agent."""
    agent = await db.agents.get("default")
    if not agent:
        raise AgentNotFoundException(
            detail="Default agent configuration is missing",
            suggested_action="Contact the administrator to set up the default agent"
        )
    return agent


@router.get("/{agent_id}", response_model=AgentResponse)
async def get_agent(agent_id: str):
    """Get a specific agent by ID."""
    agent = await db.agents.get(agent_id)
    if not agent:
        raise AgentNotFoundException(
            detail=f"Agent with ID '{agent_id}' does not exist",
            suggested_action="Please check the agent ID and try again"
        )
    return agent


@router.post("", response_model=AgentResponse, status_code=201)
async def create_agent(request: AgentCreateRequest):
    """Create a new agent."""
    agent_data = {
        "name": request.name,
        "description": request.description,
        "model": request.model,
        "permission_mode": request.permission_mode,
        "max_turns": request.max_turns,
        "system_prompt": request.system_prompt,
        "allowed_tools": request.allowed_tools,
        "skill_ids": request.skill_ids,
        "mcp_ids": request.mcp_ids,
        "working_directory": None,  # Use default from settings.agent_workspace_dir
        "enable_bash_tool": request.enable_bash_tool,
        "enable_file_tools": request.enable_file_tools,
        "enable_web_tools": request.enable_web_tools,
        "enable_tool_logging": True,
        "enable_safety_checks": True,
        "status": "active",
    }
    agent = await db.agents.put(agent_data)
    return agent


@router.put("/{agent_id}", response_model=AgentResponse)
async def update_agent(agent_id: str, request: AgentUpdateRequest):
    """Update an existing agent."""
    existing = await db.agents.get(agent_id)
    if not existing:
        raise AgentNotFoundException(
            detail=f"Agent with ID '{agent_id}' does not exist",
            suggested_action="Please check the agent ID and try again"
        )

    updates = request.model_dump(exclude_unset=True)
    agent = await db.agents.update(agent_id, updates)
    return agent


@router.delete("/{agent_id}", status_code=204)
async def delete_agent(agent_id: str):
    """Delete an agent."""
    if agent_id == "default":
        raise ValidationException(
            message="Cannot delete the default agent",
            detail="The default agent is a system resource and cannot be deleted",
            suggested_action="If you need to modify the default agent, use the update endpoint instead"
        )

    deleted = await db.agents.delete(agent_id)
    if not deleted:
        raise AgentNotFoundException(
            detail=f"Agent with ID '{agent_id}' does not exist",
            suggested_action="Please check the agent ID and try again"
        )
